### PR TITLE
Support for Base64url with case insensitive padding

### LINKED ISSUE #205: base64url decode fails on valid base64url due to capital padding
**Describe the bug**
The base64url decoder does not accept "%3D" as padding, only "%3d" (with lowercase d). As far as I can tell, uppercase is also a valid representation of the padding, and even the predominant one, judging by the google results.

**How To Reproduce**
As an example: The base64url-encoded version of "1" with uppercase D ("MQ%3D%3D") fails to decode.

**Expected behavior**
"%3D" should also be accepted as padding.

**Additional context**
This was discovered because Keycloak changes padding in the state parameter of an authorization request from "%3d" to "%3D" before passing it on to the redirect uri.

### PR DESCRIPTION
**Pull request by prince-chrismc**
resolve #205

### CODE REVIEW

File: include/jwt-cpp/base.h [hunk 1]
```diff
@@ -41,168 +46,187 @@ namespace jwt {
...
+					return lhs.count == rhs.count && lhs.length == rhs.length;
+				}
+			};
 
-				res += alphabet[(triple >> 3 * 6) & 0x3F];
-				res += alphabet[(triple >> 2 * 6) & 0x3F];
-				res += alphabet[(triple >> 1 * 6) & 0x3F];
-				res += alphabet[(triple >> 0 * 6) & 0x3F];
-			}
+			inline padding count_padding(const std::string& base, const std::initializer_list<std::string>& fills) {
```
Comment (prince-chrismc) [inline include/jwt-cpp/base.h:88]: I refactored the "count padding" to take a list of fills and measure them....

File: include/jwt-cpp/base.h [hunk 2]
```diff
@@ -41,168 +46,187 @@ namespace jwt {
...
+					res += fill;
 					break;
+				default: break;
+				}
+
+				return res;
 			}
 
-			if ((size + fill_cnt) % 4 != 0) throw std::runtime_error("Invalid input: incorrect total size");
+			inline std::string decode(const std::string& base, const alphabet::soup& alphabet,
```
Comment (prince-chrismc) [inline include/jwt-cpp/base.h:149]: I'd like to overload this with a new "case insensitive" alphabet

File: tests/BaseTest.cpp [hunk 3]
```diff
@@ -1,6 +1,48 @@
...
+	ASSERT_EQ((padding{1, 1}), jwt::base::details::count_padding("ABC!", {"~", "!"}));
+	ASSERT_EQ((padding{1, 1}), jwt::base::details::count_padding("ABC~", {"~", "!"}));
+	ASSERT_EQ((padding{3, 3}), jwt::base::details::count_padding("ABC~~!", {"~", "!"}));
+	ASSERT_EQ((padding{3, 3}), jwt::base::details::count_padding("ABC!~~", {"~", "!"}));
+	ASSERT_EQ((padding{5, 5}), jwt::base::details::count_padding("ABC~~!~~", {"~", "!"}));
+
+	ASSERT_EQ((padding{2, 6}), jwt::base::details::count_padding("MTIzNA%3d%3d", {"%3d", "%3D"}));
+	ASSERT_EQ((padding{2, 6}), jwt::base::details::count_padding("MTIzNA%3d%3D", {"%3d", "%3D"}));
+	ASSERT_EQ((padding{2, 6}), jwt::base::details::count_padding("MTIzNA%3D%3d", {"%3d", "%3D"}));
+	ASSERT_EQ((padding{2, 6}), jwt::base::details::count_padding("MTIzNA%3D%3D", {"%3d", "%3D"}));
```
Comment (prince-chrismc) [inline tests/BaseTest.cpp:32]: You can see it can matches both now

**Review by prince-chrismc — COMMENTED** [review]
I'd love a sanity check but I think I found an easy way to fix the base64url capitalization bug

**Comment by prince-chrismc** [comment]
@Thalhammer When you have a chance, I'd love a second set of eyes to double check my logic 🙏

**Review by Thalhammer — APPROVED** [review]

**Comment by Thalhammer** [comment]
Looks good to me ;)

**Pull request merged**